source/rust_verify/src/names.rs

```diff
@@ -0,0 +1,272 @@
+            }
+        }
+    }
+
+    pub(crate) fn get_impl_id<'tcx>(
+        &self,
+        tcx: TyCtxt<'tcx>,
+        def_path: DefPath,
+        num_segments: usize,
+    ) -> Option<Ident> {
+        let id = def_path_to_impl_name_id(tcx, def_path, num_segments);
+        self.impl_names.get(&id).cloned()
+    }
+}
+
+impl TypTree {
+    fn short_name(&self) -> String {
+        match self {
+            TypTree::String(s) => s.clone(),
+            TypTree::Path(p) | TypTree::Apply(p, _) => p.last().expect("path").clone(),
+            TypTree::Decorate(_, t) => t.short_name(),
+        }
+    }
+}
+
+fn typ_tree<'tcx>(tcx: TyCtxt<'tcx>, ty: &Ty<'tcx>) -> TypTree {
+    let box_rec = |ty: &Ty<'tcx>| -> Box<TypTree> { Box::new(typ_tree(tcx, ty)) };
+    match ty.kind() {
+        TyKind::Bool | TyKind::Char | TyKind::Uint(_) | TyKind::Int(_) | TyKind::Float(_) => {
+            TypTree::String(ty.to_string())
+        }
+        TyKind::Adt(AdtDef(adt), args) => {
+            let path = def_path_to_path(tcx, tcx.def_path(adt.did));
+            let mut typ_args: Vec<Box<TypTree>> = Vec::new();
+            for arg in args.iter() {
+                match arg.unpack() {
+                    rustc_middle::ty::subst::GenericArgKind::Type(t) => {
+                        typ_args.push(box_rec(&t));
+                    }
+                    _ => {}
+                }
+            }
+            TypTree::Apply(path, typ_args)
+        }
+        TyKind::Foreign(did) => TypTree::Path(def_path_to_path(tcx, tcx.def_path(*did))),
+        TyKind::Str => TypTree::String("str".to_string()),
+        TyKind::Array(t, _len) => TypTree::Decorate("array".to_string(), box_rec(t)),
+        TyKind::Slice(t) => TypTree::Decorate("slice".to_string(), box_rec(t)),
+        TyKind::RawPtr(tmut) => {
+            TypTree::Decorate(format!("raw{:?}", tmut.mutbl), box_rec(&tmut.ty))
+        }
+        TyKind::Ref(_region, t, muta) => TypTree::Decorate(format!("ref{:?}", muta), box_rec(t)),
+        TyKind::Never => TypTree::String("never".to_string()),
+        TyKind::Tuple(ts) => {
+            let path = vec!["tuple".to_string()];
+            let typ_args = ts.iter().map(|t| box_rec(&t)).collect();
+            TypTree::Apply(path, typ_args)
+        }
+        TyKind::Alias(_kind, alias) => {
+            let path = def_path_to_path(tcx, tcx.def_path(alias.def_id));
+            let mut typ_args: Vec<Box<TypTree>> = Vec::new();
+            if let Some(args) = alias.substs.try_as_type_list() {
+                for t in args.iter() {
+                    typ_args.push(box_rec(&t));
+                }
+            }
+            TypTree::Apply(path, typ_args)
+        }
+        TyKind::Param(_) => TypTree::String(ty.to_string()),
+        TyKind::Bound(..) => TypTree::String(ty.to_string()),
+
+        TyKind::FnDef(..) => TypTree::String("fndef".to_string()),
+        TyKind::FnPtr(..) => TypTree::String("fnptr".to_string()),
```

Comment:
why do FnDef and FnPtr not include type arguments?

Reply:
These should probably be `unsupported_err!`, as in `mid_ty_to_vir_ghost`.  I just have to set up the spans and `Result` return types so that these errors would get reported properly.

Reply:
Ok, I updated the pull request to use `unsupported_err!`